src/bin/pgcopydb/ld_replay.c

```diff
@@ -249,3 +257,300 @@ stream_replay_line(void *ctx, const char *line, bool *stop)
+			 * start buffering from the current transaction.
+			 */
+			log_debug("Received %s when transaction is already "
+					  "in buffering mode, ignoring.", line);
+			fclose(replayCtx->txnBuffer);
+
+			/* Remove symlink to the replay buffer */
+			(void) unlink_file(context->paths.txnlatestfile);
+		}
+
+		sformat(replayCtx->currentTxnFileName, MAXPGPATH, "%s/%d.sql",
+				context->paths.dir, metadata.xid);
+
+		/*
+		 * Open current transaction file in w+ mode to truncate if the file
+		 * already exists.
+		 */
+		replayCtx->txnBuffer = fopen_with_umask(replayCtx->currentTxnFileName,
+												"w+",
+												O_RDWR | O_TRUNC | O_CREAT,
+												0644);
+		if (replayCtx->txnBuffer == NULL)
+		{
+			log_error("Failed to open transaction buffer file \"%s\": %m",
+					  replayCtx->currentTxnFileName);
+			return false;
+		}
+
+		/* Create as a symlink to the current transcation file */
+		if (!create_symbolic_link(replayCtx->currentTxnFileName,
+								  context->paths.txnlatestfile))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		fformat(replayCtx->txnBuffer, "%s\n", line);
+
+		replayCtx->isTxnBuffering = true;
+	}
+	else if (metadata.action == STREAM_ACTION_COMMIT)
+	{
+		if (replayCtx->skipTxnBuffering)
+		{
+			return stream_replay_line(ctx, line, stop);
+		}
+
+		if (!replayCtx->isTxnBuffering)
+		{
+			/*
+			 * When the follow switches from prefetch to replay mode, we
+			 * call stream_transform_stream which might stream the partially
+			 * written transaction created during the prefetch mode.
+			 *
+			 * For example, lets consider the partially written C.sql file,
+			 *
+			 * COMMIT -- {"xid": 999, "lsn": "0/1230"};
+			 * BEGIN -- {"xid": 1000, "commitLSN": "0/1234"};
+			 * INSERT INTO C VALUES (1);
+			 * INSERT INTO C VALUES (2);
+			 * KEEPALIVE;
+			 *
+			 * After switching to the replay mode, logical decoding will resume
+			 * from consistent point which again starts with a valid transcation
+			 * block.
+			 *
+			 * Lets assume the following contents streamed from the transform
+			 * to catchup process in UNIX PIPE after switching to replay mode,
+			 *
+			 * BEGIN -- {"xid": 999, "commitLSN": "0/1230"};
+			 * INSERT INTO A VALUES (1);
+			 * INSERT INTO A VALUES (2);
+			 * COMMIT -- {"xid": 999, "lsn": "0/1230"};
+			 * BEGIN -- {"xid": 1000, "commitLSN": "0/1234"};
+			 * INSERT INTO C VALUES (1);
+			 * INSERT INTO C VALUES (2);
+			 * COMMIT -- {"xid": 1000, "lsn": "0/1234"};
+			 *
+			 * Contents of C.sql will be streamed by stream_transform_stream
+			 * and followed by the contents from UNIX PIPE. Since both of the
+			 * contents are streamed one after the other, the second block
+			 * contains the full content of previously written
+			 * transaction. So, we can skip previous transaction buffer and
+			 * start buffering from the current transaction.
+			 */
+			log_debug("Received %s when transaction is not "
+					  "in buffering mode, ignoring.", line);
```

Comment:
Please rewrite the message so that it ends with `": %s", line`. This makes it easier to process the logs later.